src/responder/secrets/providers.c

```diff
@@ -42,32 +43,190 @@ static int sec_map_url_to_user_path(struct sec_req_ctx *secreq,
         return ENOMEM;
     }
 
-    DEBUG(SSSDBG_TRACE_LIBS, "User-specific path is [%s]\n", *mapped_path);
+    DEBUG(SSSDBG_TRACE_LIBS,
+          "User-specific secrets path is [%s]\n", *mapped_path);
     return EOK;
 }
 
-int sec_req_routing(TALLOC_CTX *mem_ctx, struct sec_req_ctx *secreq,
-                    struct provider_handle **handle)
+/* The base for storing ccaches is:
+ *  http://localhost/kcm/persistent/$uid
+ *
+ * Under $base, there are two containers:
+ *  /default    - stores the name of the default ccache for this UID
+ *  /ccache     - stores the ccaches
+ *  /ntlm       - stores NTLM creds [Not implement yet]
+ *
+ * Each ccache has a name and an UUID. On the secrets level, the 'secret'
+ * is a concatenation of the stringified UUID and the name.
+ */
+
+static errno_t check_kcm_uid(const char **relpath,
+                             uid_t *_uid)
+{
+    const int max_uid_len = 12;
+    uint32_t uid;
+    char *endptr;
+    const char *str_end;
+    size_t str_uid_len;
+    char str_uid[max_uid_len];
+
+    str_end = strchr(*relpath, '/');
+    if (str_end == NULL) {
+        return EINVAL;
+    }
+    str_uid_len = str_end - *relpath;
+
+    if (str_uid_len >= max_uid_len) {
+        /* Just make sure we don't overflow the buffer, if the number
+         * is larger than UINT32_MAX is checked with strtouint32_t
+         */
+        return E2BIG;
+    }
+    strncpy(str_uid, *relpath, str_uid_len);
+    str_uid[str_uid_len] = '\0';
+
+    errno = 0;
+    uid = strtouint32(str_uid, &endptr, 10);
+    if (errno != 0) {
+        return errno;
+    } else if (*endptr) {
+        return EINVAL;
+    }
+
+    *_uid = uid;
+    *relpath = str_end;
+    return EOK;
+}
+
+/*
+ * Internally, we map the incoming path to:
+ * SEC_KCM_PFX/$session/$impersonated_uid/$relpath
+ *
```

Comment:
this was fixed in the previous iteration already. now only the peer credentials are checked and the patch is forwarded to the provider